apiserver/facades/agent/provisioner/provisioner.go

```diff
@@ -507,6 +507,10 @@ func (api *ProvisionerAPI) AvailabilityZone(ctx context.Context, args params.Ent
 			continue
 		}
 		hc, err := api.machineService.HardwareCharacteristics(ctx, machineUUID)
+		if errors.Is(err, machineerrors.NotProvisioned) {
+			result.Results[i].Error = apiservererrors.ServerError(machineerrors.NotProvisioned)
+			continue
+		}
```

Comment:
@SimonRichardson I updated the returned error on `HardwareCharacteristics()` when `ErrNoRows` to be a `NotProvisioned`, because we are selecting from the `machine_cloud_instance` table and the error should be the same as the one we already returned from `InstanceID()` when no rows were found.
I added this specific error check on the call sites.